File: notes.swift
// notes.swift
import Foundation

let noteFreqs: [String: Double] = [
    "C": 261.63, "C#": 277.18, "Db": 277.18,
    "D": 293.66, "D#": 311.13, "Eb": 311.13,
    "E": 329.63,
    "F": 349.23, "F#": 369.99, "Gb": 369.99,
    "G": 392.00, "G#": 415.30, "Ab": 415.30,
    "A": 440.00, "A#": 466.16, "Bb": 466.16,
    "B": 493.88
]

func parseNote(_ note: String) -> (freq: Double?, parsed: String?) {
    let n = note.trimmingCharacters(in: .whitespacesAndNewlines)
    if n.isEmpty { return (nil, nil) }
    var letter = ""
    var octave = 4
    var i = n.startIndex
    while i < n.endIndex && (n[i].isLetter || n[i] == "#") {
        letter.append(n[i])
        i = n.index(after: i)
    }
    if i < n.endIndex && n[i].isNumber {
        let octStr = String(n[i...])
        octave = Int(octStr) ?? 4
    }
    if let base = noteFreqs[letter] {
        let freq = base * pow(2.0, Double(octave - 4))
        return (freq, n)
    }
    return (nil, nil)
}

func playFreq(_ freq: Double?) {
    guard let f = freq, f > 0 else { return }
    #if os(macOS)
    print("  (would play \(String(format: "%.1f", f)) Hz on macOS)")
    #elseif os(Windows)
    print("  (would play \(String(format: "%.1f", f)) Hz on Windows)")
    #else
    // Linux: try beep command
    let process = Process()
    process.executableURL = URL(fileURLWithPath: "/usr/bin/env")
    process.arguments = ["beep", "-f", String(Int(f)), "-l", "300"]
    process.standardOutput = FileHandle.nullDevice
    process.standardError = FileHandle.nullDevice
    try? process.run()
    #endif
}

func playNotes(_ notes: [String]) {
    if notes.isEmpty {
        print("No notes to play.")
        return
    }
    print("Playing melody...")
    for n in notes {
        let (freq, parsed) = parseNote(n)
        guard let f = freq, let p = parsed else {
            print("  Skipping unknown note: \(n)")
            continue
        }
        print("  \(p) (\(String(format: "%.1f", f)) Hz)")
        playFreq(f)
        Thread.sleep(forTimeInterval: 0.3)
    }
}

func main() {
    var notes: [String] = []
    print("=== Notation Editor ===")
    while true {
        print("\n1. Add note")
        print("2. Remove note")
        print("3. List notes")
        print("4. Play melody")
        print("5. Save to file")
        print("6. Load from file")
        print("7. Exit")
        print("Choose: ", terminator: "")
        guard let choice = readLine()?.trimmingCharacters(in: .whitespacesAndNewlines) else { continue }
        switch choice {
        case "1":
            print("Enter note (e.g., C4, D#4): ", terminator: "")
            guard let note = readLine()?.trimmingCharacters(in: .whitespacesAndNewlines) else { break }
            if !note.isEmpty {
                notes.append(note)
                print("Added: \(note)")
            } else {
                print("Invalid note.")
            }
        case "2":
            if notes.isEmpty {
                print("No notes to remove.")
                break
            }
            print("Index to remove: ", terminator: "")
            guard let idxStr = readLine(), let idx = Int(idxStr.trimmingCharacters(in: .whitespacesAndNewlines)) else {
                print("Invalid number.")
                break
            }
            if idx >= 1 && idx <= notes.count {
                let removed = notes.remove(at: idx-1)
                print("Removed: \(removed)")
            } else {
                print("Invalid index.")
            }
        case "3":
            if notes.isEmpty {
                print("No notes.")
            } else {
                for (i, n) in notes.enumerated() {
                    print("[\(i+1)] \(n)")
                }
            }
        case "4":
            playNotes(notes)
        case "5":
            print("Filename: ", terminator: "")
            guard let fname = readLine()?.trimmingCharacters(in: .whitespacesAndNewlines), !fname.isEmpty else { break }
            let content = notes.joined(separator: "\n")
            do {
                try content.write(toFile: fname, atomically: true, encoding: .utf8)
                print("Saved to \(fname)")
            } catch {
                print("Error saving: \(error)")
            }
        case "6":
            print("Filename: ", terminator: "")
            guard let fname = readLine()?.trimmingCharacters(in: .whitespacesAndNewlines), !fname.isEmpty else { break }
            do {
                let content = try String(contentsOfFile: fname, encoding: .utf8)
                let loaded = content.split(separator: "\n").map { String($0).trimmingCharacters(in: .whitespacesAndNewlines) }.filter { !$0.isEmpty }
                notes = loaded
                print("Loaded \(notes.count) notes from \(fname)")
            } catch {
                print("File not found.")
            }
        case "7":
            print("Goodbye!")
            return
        default:
            print("Invalid choice.")
        }
    }
}

main()

File: Notes.java
// Notes.java
import java.io.*;
import java.util.*;
import java.lang.Math;

public class Notes {
    private static List<String> notes = new ArrayList<>();
    private static final Map<String, Double> NOTE_FREQS = new HashMap<>();
    static {
        NOTE_FREQS.put("C", 261.63); NOTE_FREQS.put("C#", 277.18); NOTE_FREQS.put("Db", 277.18);
        NOTE_FREQS.put("D", 293.66); NOTE_FREQS.put("D#", 311.13); NOTE_FREQS.put("Eb", 311.13);
        NOTE_FREQS.put("E", 329.63);
        NOTE_FREQS.put("F", 349.23); NOTE_FREQS.put("F#", 369.99); NOTE_FREQS.put("Gb", 369.99);
        NOTE_FREQS.put("G", 392.00); NOTE_FREQS.put("G#", 415.30); NOTE_FREQS.put("Ab", 415.30);
        NOTE_FREQS.put("A", 440.00); NOTE_FREQS.put("A#", 466.16); NOTE_FREQS.put("Bb", 466.16);
        NOTE_FREQS.put("B", 493.88);
    }

    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        System.out.println("=== Notation Editor ===");
        while (true) {
            System.out.println("\n1. Add note");
            System.out.println("2. Remove note");
            System.out.println("3. List notes");
            System.out.println("4. Play melody");
            System.out.println("5. Save to file");
            System.out.println("6. Load from file");
            System.out.println("7. Exit");
            System.out.print("Choose: ");
            String choice = reader.readLine().trim();
            switch (choice) {
                case "1":
                    System.out.print("Enter note (e.g., C4, D#4): ");
                    String note = reader.readLine().trim();
                    if (!note.isEmpty()) {
                        notes.add(note);
                        System.out.println("Added: " + note);
                    } else System.out.println("Invalid note.");
                    break;
                case "2":
                    if (notes.isEmpty()) {
                        System.out.println("No notes to remove.");
                        break;
                    }
                    System.out.print("Index to remove: ");
                    try {
                        int idx = Integer.parseInt(reader.readLine().trim());
                        if (idx >= 1 && idx <= notes.size()) {
                            String removed = notes.remove(idx - 1);
                            System.out.println("Removed: " + removed);
                        } else System.out.println("Invalid index.");
                    } catch (NumberFormatException e) {
                        System.out.println("Invalid number.");
                    }
                    break;
                case "3":
                    if (notes.isEmpty()) System.out.println("No notes.");
                    else {
                        for (int i = 0; i < notes.size(); i++)
                            System.out.println("[" + (i+1) + "] " + notes.get(i));
                    }
                    break;
                case "4":
                    playMelody();
                    break;
                case "5":
                    System.out.print("Filename: ");
                    String fname = reader.readLine().trim();
                    if (!fname.isEmpty()) {
                        try (PrintWriter pw = new PrintWriter(fname)) {
                            for (String n : notes) pw.println(n);
                        }
                        System.out.println("Saved to " + fname);
                    }
                    break;
                case "6":
                    System.out.print("Filename: ");
                    fname = reader.readLine().trim();
                    try (BufferedReader br = new BufferedReader(new FileReader(fname))) {
                        List<String> loaded = new ArrayList<>();
                        String line;
                        while ((line = br.readLine()) != null) {
                            line = line.trim();
                            if (!line.isEmpty()) loaded.add(line);
                        }
                        notes = loaded;
                        System.out.println("Loaded " + notes.size() + " notes from " + fname);
                    } catch (FileNotFoundException e) {
                        System.out.println("File not found.");
                    }
                    break;
                case "7":
                    System.out.println("Goodbye!");
                    return;
                default:
                    System.out.println("Invalid choice.");
            }
        }
    }

    private static void playMelody() {
        if (notes.isEmpty()) {
            System.out.println("No notes to play.");
            return;
        }
        System.out.println("Playing melody...");
        for (String n : notes) {
            double[] freq = parseNote(n);
            if (freq == null) {
                System.out.println("  Skipping unknown note: " + n);
                continue;
            }
            System.out.printf("  %s (%.1f Hz)%n", n, freq[0]);
            // Attempt to beep using Java's Toolkit (not available) – print only.
            // Java doesn't have a simple beep without external libs.
        }
    }

    private static double[] parseNote(String note) {
        note = note.trim();
        if (note.isEmpty()) return null;
        String letter = "";
        int octave = 4;
        int i = 0;
        while (i < note.length() && (Character.isLetter(note.charAt(i)) || note.charAt(i) == '#')) {
            letter += note.charAt(i);
            i++;
        }
        if (i < note.length() && Character.isDigit(note.charAt(i))) {
            octave = Integer.parseInt(note.substring(i));
        }
        if (NOTE_FREQS.containsKey(letter)) {
            double base = NOTE_FREQS.get(letter);
            double freq = base * Math.pow(2, octave - 4);
            return new double[]{freq};
        }
        return null;
    }
}
